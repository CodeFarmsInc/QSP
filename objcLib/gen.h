#ifndef ZZ_CODEGEN_INCLUDE
#define ZZ_CODEGEN_INCLUDE

#include "lib/general.h"

class Generator;
class Organization;
class Participant;
class OrgType;
class PartType;
class Iterator;
class FixedType;
class ClassLink;
class ApplClass;

#include "tmp/orgs.h"
#include "tmp/orgs1.h"
#include "tmp/parts.h"
#include "tmp/parts1.h"
#include "tmp/parts2.h"
#include "tmp/orgName.h"
#include "tmp/partName.h"
#include "tmp/types.h"
#include "tmp/types1.h"
#include "tmp/partTypes.h"
#include "tmp/partTypes1.h"
#include "tmp/partTypes2.h"
#include "tmp/inherits.h"
#include "tmp/iters.h"
#include "tmp/iters1.h"
#include "tmp/orgTypeName.h"
#include "tmp/fileName.h"
#include "tmp/partTypeName.h"
#include "tmp/iterName.h"
#include "tmp/inhName.h"
#include "tmp/fixedName.h"
#include "tmp/inhMatch.h"
#include "tmp/inhFixed.h"
#include "tmp/fixedTypes.h"
#include "tmp/fixedTypes1.h"
#include "tmp/myType.h"
#include "tmp/myPartType.h"
#include "tmp/toBase.h"
#include "tmp/classes.h"
#include "tmp/classes1.h"
#include "tmp/dependsOn.h"
#include "tmp/dependsOn1.h"
#include "tmp/usedBy.h"
#include "tmp/usedBy1.h"
#include "tmp/usedBy2.h"
#include "tmp/cName.h"

// -----------------------------------------------
typedef class orgs_LinkedList2 orgs;
typedef class parts_Aggregate2 parts;
typedef class orgName_Name orgName;
typedef class partName_Name partName;
typedef class types_LinkedList2 types;
typedef class partTypes_Aggregate2 partTypes;
typedef class inherits_SingleLink inherits;
typedef class iters_LinkedList1 iters;
typedef class orgTypeName_Name orgTypeName;
typedef class fileName_Name fileName;
typedef class partTypeName_Name partTypeName;
typedef class iterName_Name iterName;
typedef class inhName_Name inhName;
typedef class fixedName_Name fixedName;
typedef class inhMatch_SingleLink inhMatch;
typedef class inhFixed_SingleLink inhFixed;
typedef class fixedTypes_LinkedList1 fixedTypes;
typedef class myType_SingleLink myType;
typedef class myPartType_SingleLink myPartType;
typedef class toBase_SingleLink toBase;
typedef class classes_Hash classes;
typedef class dependsOn_LinkedList1 dependsOn;
typedef class usedBy_Aggregate2 usedBy;
typedef class cName_Name cName;
// -----------------------------------------------


class ZZ_Generator {
public:
    orgs_LinkedList2Parent ZZorgs;
    types_LinkedList2Parent ZZtypes;
    classes_HashHolder ZZclasses;
};

class ZZ_Organization {
public:
    orgs_LinkedList2Child ZZorgs;
    parts_Aggregate2Parent ZZparts;
    orgName_NameParent ZZorgName;
    myType_SingleLinkParent ZZmyType;
};

class ZZ_Participant {
public:
    parts_Aggregate2Child ZZparts;
    partName_NameParent ZZpartName;
    myPartType_SingleLinkParent ZZmyPartType;
    usedBy_Aggregate2Child ZZusedBy;
};

class ZZ_OrgType {
public:
    types_LinkedList2Child ZZtypes;
    partTypes_Aggregate2Parent ZZpartTypes;
    inherits_SingleLinkParent ZZinherits;
    iters_LinkedList1Parent ZZiters;
    orgTypeName_NameParent ZZorgTypeName;
    fileName_NameParent ZZfileName;
    fixedTypes_LinkedList1Parent ZZfixedTypes;
};

class ZZ_PartType {
public:
    partTypes_Aggregate2Child ZZpartTypes;
    partTypeName_NameParent ZZpartTypeName;
    inhName_NameParent ZZinhName;
    inhMatch_SingleLinkParent ZZinhMatch;
};

class ZZ_Iterator {
public:
    iters_LinkedList1Child ZZiters;
    iterName_NameParent ZZiterName;
};

class ZZ_FixedType {
public:
    fixedName_NameParent ZZfixedName;
    inhFixed_SingleLinkParent ZZinhFixed;
    fixedTypes_LinkedList1Child ZZfixedTypes;
};

class ZZ_ClassLink {
public:
    toBase_SingleLinkParent ZZtoBase;
    dependsOn_LinkedList1Child ZZdependsOn;
};

class ZZ_ApplClass {
public:
    classes_HashElement ZZclasses;
    dependsOn_LinkedList1Parent ZZdependsOn;
    usedBy_Aggregate2Parent ZZusedBy;
    cName_NameParent ZZcName;
};

typedef orgs_LinkedList2Iterator orgs_Iterator;
typedef parts_Aggregate2Iterator parts_Iterator;
typedef types_LinkedList2Iterator types_Iterator;
typedef partTypes_Aggregate2Iterator partTypes_Iterator;
typedef iters_LinkedList1Iterator iters_Iterator;
typedef fixedTypes_LinkedList1Iterator fixedTypes_Iterator;
typedef classes_HashIterator classes_Iterator;
typedef dependsOn_LinkedList1Iterator dependsOn_Iterator;
typedef usedBy_Aggregate2Iterator usedBy_Iterator;

#endif // ZZ_CODEGEN_INCLUDE
