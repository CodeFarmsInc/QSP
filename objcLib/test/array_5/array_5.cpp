// -------------------------------------------------------------------------
// This is a test of accessing bits of Array - coded in the DOL style.
// Since this library has a special class for this purpose (BitArray), 
// this approach is not recommended unless you are converting old DOL code.
// The results should be identical with the results of test59 from
// the DOL regression suite, and the results
//  NOTE: The old style where array of int's or char's is treated as
//        an array of bits:                  BinArray<Root,int> bitArr; 
//        The new style, array of bits:      BitArray<Root> bitArr;

// (1) We initial array of integers by some arbitrary numbers, and then
// see that we can access the bits properly.
// (2) We change some of the bits and see what effect it has on the integers.
// (3) We set bits outside of the initial array size, and see that we still
//     can access them properly.
//
// WARNING: Results may be different for different hardware architectures,
//          On a PC, integer 5 is represented as 00...101
//                   integer -1               as 11...110
// --------------------------------------------------------------------------

#include <stdio.h>
#include "gen.h"

class Root {     // class under which to store the array
public:
    ZZ_Root ZZds;
    void prt(); // print the current bit array
};


int main(){
    Root *rp; int i,k,sz,cap,incr,*iPtr;

    rp=new Root;
    bitArr.form(rp,3,-2); // initial size=3, increase 2x when needed
    cap=bitArr.capacity(rp);
    sz=bitArr.size(rp);
    incr=bitArr.increment(rp);
    printf("starting: capacity=%d size=%d incr=%d\n",cap,sz,incr);

    // test (1): set some numbers
    printf("TEST 1:\n");
    iPtr=bitArr.ind(rp,0); *iPtr= 34;   // bits 1 and 5
    iPtr=bitArr.ind(rp,1); *iPtr=257;   // bits 32 and 40
    iPtr=bitArr.ind(rp,2); *iPtr= -1;   // all bits from 64 to 95

    // see that the bits are interpreted properly
    rp->prt();

    unsigned char * p=(unsigned char*)bitArr.ind(rp,0);
    for(i=0; i<12; i++, p++)printf("%o ",*p); printf("\n");

    // test (2): change some bits, see the result on the integers
    printf("\nTEST 2:\n");
    printf("\n");
    bitArr.setBit(rp,0,1);
    bitArr.setBit(rp,1,1); // no change really
    bitArr.setBit(rp,40,0);
    bitArr.setBit(rp,94,0);
    bitArr.setBit(rp,95,0);

    rp->prt();

    // test(3): set bits outside of the original range, print values
    printf("\nTEST 3:\n");
    bitArr.setBit(rp,500,1);
    bitArr.setBit(rp,501,0);
    bitArr.setBit(rp,502,1);
    printf("%d",bitArr.getBit(rp,500));
    printf("%d",bitArr.getBit(rp,501));
    printf("%d",bitArr.getBit(rp,502));

    cap=bitArr.capacity(rp);
    sz=bitArr.size(rp);
    incr=bitArr.increment(rp);
    printf(" capacity=%d size=%d incr=%d\n",cap,sz,incr);

    rp->prt();
    bitArr.free(rp);
    delete rp;
    return 0;
}

// ----------------------------------------------------
// Print the current bit array
// ----------------------------------------------------
void Root::prt(){
    int i,k,sz,v;

    sz=bitArr.size(this);
    for(i=0; i<sz; i++){
        for(k=0; k<32; k++){
            v=bitArr.getBit(this,i*32+k);
            printf("%d",v);
        }
        printf("\n");
    }
    printf("\n");
}

#include "gen.cpp"

