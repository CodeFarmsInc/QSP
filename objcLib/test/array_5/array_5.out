starting: capacity=3 size=0 incr=-2
TEST 1:
01000100000000000000000000000000
10000000100000000000000000000000
11111111111111111111111111111111

42 0 0 0 1 1 0 0 377 377 377 377 

TEST 2:

11000100000000000000000000000000
10000000000000000000000000000000
11111111111111111111111111111100


TEST 3:
101 capacity=24 size=16 incr=-2
11000100000000000000000000000000
10000000000000000000000000000000
11111111111111111111111111111100
00000000000000000000000000000000
00000000000000000000000000000000
00000000000000000000000000000000
00000000000000000000000000000000
00000000000000000000000000000000
00000000000000000000000000000000
00000000000000000000000000000000
00000000000000000000000000000000
00000000000000000000000000000000
00000000000000000000000000000000
00000000000000000000000000000000
00000000000000000000000000000000
00000000000000000000101000000000

